### ISSUE TITLE
Wayland compatibility

### ISSUE THREAD

**Issue by dr-ni** [onboard-osk/onboard#3]
What are the next steps for wayland?

**Comment by dr-ni** [comment]
some xkb infos:
https://unix.stackexchange.com/questions/309580/does-wayland-use-xkb-for-keyboard-layouts
https://hack.org/~mc/writings/xkb.html
https://unix.stackexchange.com/questions/521136/key-remapping-for-special-characters-umlauts-in-wayland

**Comment by dr-ni** [comment]
keyboard grab:
https://wayland.app/protocols/xwayland-keyboard-grab-unstable-v1

**Comment by dr-ni** [comment]
simple sample:
https://stackoverflow.com/questions/76698957/wayland-minimal-keyboard-input-example

**Comment by Rmano** [comment]
Maybe this https://github.com/jjsullivan5196/wvkbd can have some hints on how to proceed? (Sorry, this is too far away from my skillset to be able to help more...)

**Comment by gnu-like** [comment]
Did you ask AI to help ?

**Comment by dr-ni** [comment]
Yes, any help would be great :-)

**Comment by gnu-like** [comment]
> Yes, any help would be great :-)

I hope you find and fix it for wayland display server

**Comment by Hephaestus985** [comment]
Any updates?

**Comment by dr-ni** [comment]
no, unfortunateley no wayland improvements, until now. We still haven't found anybody to help us until now

**Comment by Lg42tv** [comment]
I managed to run it on wayland using these steps:
https://discuss.kde.org/t/plasma-6-and-wayland-no-on-screen-keyboard-working/17799/47

**Comment by Hephaestus985** [comment]
I also managed to get Onboard working on wayland Kubuntu 25.04

Original post by [INVICTRA](https://discuss.kde.org/u/INVICTRA)  [https://discuss.kde.org/t/plasma-6-and-wayland-no-on-screen-keyboard-working/17799/47](url)

1 Edit the shortcut in the menu. Within the KDE Menu Editor look for the Environment variables field and add “GDK_BACKEND=x11”.

<img width="2572" height="1123" alt="Image" src="https://github.com/user-attachments/assets/0ec583ba-f9b2-4930-8b56-48cc5693cfe9" />

2 Go to Onboard preferences page. Under Keyboard – >Advanced set:
Input Options → Input event source: GTK
Key-stroke Generation → Key-stroke generator: uinput

<img width="1919" height="1086" alt="Image" src="https://github.com/user-attachments/assets/89b91a3a-0c15-4f31-81bd-29b953ff6f74" />

**Comment by Hephaestus985** [comment]
Only the upper case on right-click option is not working now.

The letter does show a capital letter on onboard but a small letter is outputted to the text field.

**Comment by lgoio** [comment]
## Problem with Wayland

Wayland has much stricter security rules than X11.
Programs are **no longer allowed to generate keyboard, mouse, or focus events on their own**,
as tools like `onboard`, `xdotool`, or `xprop` could under X11.

Instead, **input must be routed through the window manager**,
meaning each compositor requires **a dedicated implementation**.

Examples:

- `wvkbd` for **wlroots-based** compositors  
- `gjs-osk` as a **GNOME Shell extension** for the Mutter compositor
- `Maliit` for **KDE Plasma** (KWin Wayland compositor)

As a result, implementing on-screen keyboard support for Wayland is **significantly more complex** than on X11.

---

## Workaround with `GDK_BACKEND=x11`

One workaround is to **force GTK to use the X11 backend** by setting the following environment variable:

```bash
GDK_BACKEND=x11
```

However, for this to work correctly, it must be set **system-wide** – for example in `/etc/profile` or a file in `/etc/profile.d/`.

If it's only set **locally** (e.g. in a terminal or launcher), **Onboard won't function properly** under Wayland:

- It cannot send input to Wayland applications  
- It cannot detect editable text fields (no AT-SPI events)  
- It won't auto-show when a text field is focused

---

## Conclusion

This workaround **bypasses Wayland’s entire security architecture**.

> **Wouldn’t it make more sense to switch to Xorg – or to an on-screen keyboard with native support for the Wayland compositor?**

**Comment by Hephaestus985** [comment]
> This workaround bypasses Wayland’s entire security architecture.

Yes, that is true :(

> Wouldn’t it make more sense to switch to Xorg?

No, I can't even start an X11 session on Kubuntu 25.04 without crashing my PC. The default Wayland session does work without problems, and I am sure X11 support will become worse in the future

> Or to an on-screen keyboard with native support for the Wayland compositor?

Yes, this would be ideal, but I only found Maliit, which did not work for me

**Comment by Hephaestus985** [comment]
Maybe the Wayland text input method of Maliit can be added to Onboard?
[https://github.com/maliit](
https://github.com/maliit)

**Comment by lgoio** [comment]
Maliit isn’t a generic OSK backend either — it’s just **another dedicated input-method implementation for a specific Wayland compositor**, in this case **KWin (KDE Plasma)**.  
So just like `wvkbd` for wlroots and `gjs-osk` for Mutter, **Maliit is KWin’s input method**.

I’ve added **Maliit** to the list of dedicated implementations in the message above.

**Comment by Hephaestus985** [comment]
Ah, in that case, it would still be nice if at least the input methods for KDE (Maliit, KWin Wayland compositor) and GNOME (gjs-osk, Mutter compositor) could be added.

It is not a complete solution, but it would be better than nothing.

Maybe the code for Maliit and gjs-osk can serve as inspiration on how to do this.

**Comment by dr-ni** [comment]
If anybody wants some starting advice:
see:
https://github.com/dr-ni/onboard/blob/main/Onboard/osk/osk_virtkey_wayland.c
and:

[wayland-kb.c.txt](https://github.com/user-attachments/files/21559227/wayland-kb.c.txt)

[Keyboard input - The Wayland Protocol.pdf](https://github.com/user-attachments/files/21559123/Keyboard.input.-.The.Wayland.Protocol.pdf)

**Comment by rab0171610** [comment]
Setting GDK_BACKEND=x11 in the KDE start menu entries, desktop files, or from the command line works on Gentoo -- I can type in both Wayland and Xwayland applications.  I also found that it works in KDE when using the latest EndeavourOS.  I cannot get it to work in Wayland on Debian 13 or the latest Kubuntu (25.10) or KDE Neon.  All of those distros are also using KDE 6.  As suggested above setting GDK_BACKEND=x11 globally via /etc/profile or other mechanisms did not have any effect.  I tried everything I could think of to get it to work in those three distros in KDE and nothing changed.  It works in applications using Xwayland, but not those running natively in Wayland. 

For example, in Debian, using  'GDK_BACKEND=x11 onboard' works in Synaptic Package Manger (uses Xwayland by default) and not in Thunderbird or Firefox (uses Wayland backend).  I confirmed this with Xeyes to be sure.   I compared environment variables, tried unsetting the DISPLAY variable,  etc to no avail.  I could not get it to work as it does in Gentoo (and EndeavourOS).  I made sure that the everything was set the same in KDE SystemSettings>Security & Privacy>Legacy X11 App Support on all distros.  For the record, I have "Allow legacy X11 apps to read keystrokes typed in all apps: Always" and have checked (and toggled) "additionally include mouse buttons" and "Allow without asking for permission" but there was not difference in the results.
I also tried every combination of settings mentioned above: 
    " Onboard preferences page. Under Keyboard – >Advanced set:
     Input Options → Input event source: GTK
     Key-stroke Generation → Key-stroke generator: uinput"
It did not change the outcome.  I have this set to the defaults xinput and auto on Gentoo.  The settings appear to have no affect on the other distros.  Maybe there is a security setting on those distros with regards to Wayland apps?  Maybe the implementation of Xwayland is compiled with different options in those distros?  Maybe there is security or polkit type rule in a config file in /etc in those distros? For the record, SeLinux and Apparmor were both disabled in the tests.      
Additonally, I tried installing the version of onboard in the repos for all respective distros as well as compiling and installing the latest release, v1.4.3-7 here and the behavior was the same.
If anyone has any suggestions or understanding of why it works (Onboard can type in both Wayland and Xwayland apps) in EndeavourOS and Gentoo using GDK_BACKEND=x11 but not any other distros I have tried I would be interested in your thoughts or suggestions.

**Comment by Lg42tv** [comment]
Change the key-stroke generator to XTest

**Comment by rab0171610** [comment]
Thanks for the suggestion.  Maybe I wasn't clear above.
To clarify, I toggled every possible combination of all of the settings under  Onboard Settings– > Keyboard – >Advanced :
Input Options → Input event source: 
AND
Key-stroke Generation 
including XTest.  As I mentioned previously,, it did not change the outcome.  I tried every combination of settings for Input Options and Key-stroke Generation.  Regardless of what was set, onboard still only worked in Xwayland apps but not Wayland apps.  The exception was EndeavourOS.  
But just to confirm , I did just try your suggestion again in Kubuntu (25.10) and it did not did not have any effect.

**Comment by billyswong** [comment]
> Wayland has much stricter security rules than X11. Programs are **no longer allowed to generate keyboard, mouse, or focus events on their own**, as tools like `onboard`, `xdotool`, or `xprop` could under X11.
> 
> Instead, **input must be routed through the window manager**, meaning each compositor requires **a dedicated implementation**.

There is a software project called `ydotool` that provides xdotool functions but works also under Wayland, without any need for dedicated implementation per compositor.
https://github.com/ReimuNotMoe/ydotool

Could we learn from them and implement Onboard in a similar manner?

**Comment by AlexP11223** [comment]
> There is a software project called `ydotool`

Also `dotool` and `wtype`.

**Comment by AlexP11223** [comment]
Hello to everyone here. :)

In case someone needs Onboard working on Wayland right now, here's my attempt to add this. It seems to work quite well on KDE and GNOME, and also working in limited ways on other compositors. 

https://github.com/onboard-osk/onboard/pull/49

**Comment by rab0171610** [comment]
I wanted to add an update on a comment I posted above on on Oct 28, 2025.  First, using 'GDK_BACKEND=x11 onboard' is an acceptable workaround for me.  For my personal use cases on my personal desktops, I do not feel that additional security measures in Wayland are addressing problems that I have.  X11 level security is acceptable to me.   I accept the responsibility for that.

I was able to get Onboard working in Debian Stable (Trixie) by using the uinput method, but it did take some configuration.
The steps I followed were found here:
https://blog.hollen9.com/en/post/deb13k6wlonb/
A big thank you to the person that posted those helpful instructions.

For Gentoo, a recent update to Plasma 6.6 rendered Onboard not working when running under Xwayland (using the 'GDK_BACKEND=x11 onboard' method) .  I began to have the same issue I was experiencing in Debian, Onboard worked only in Xwayland windows but not native Wayland programs.  While it was not apparent what was wrong, I assumed (incorrectly as it turns out) that it was due to KDE leaning towards stricter Wayland security.  Important to note,  I was not able to get the uinput method to work either.  
It turned out that this comment from the Developer of Vboard (mdev588) led me to a solution:

[Not working on KDE plasma 6.6 (Arch Linux Wayland)]
#29
https://github.com/mdev588/vboard/issues/29#issuecomment-3937865676

[mdev588](https://github.com/mdev588)
[on Feb 20, 2026](https://github.com/mdev588/vboard/issues/29#issuecomment-3937574419)
     "Mdev588 on Feb 20:

     It seems kde started ignoring this line of code (bug?)

     self.set_accept_focus(False)

     A current workaround is manually setting vboard to be unfocusable by right clicking titlebar then chose 'configure window 
     settings' then 'Add Property' then select 'Accept focus' and set it to No."

It seems that in more recent versions of KDE Plasma, Onboard was now getting the focus i.e. accepting keystrokes instead of the window where you would expect to be typing (where the cursor is).  There were no visual cues to indicate this change in behavior so it was not obvious.      

In System Settings->Window Management->Window Rules, I created a window rule for Onboard setting Accept Focus (FORCE) to 'No': 

<img width="699" height="299" alt="Image" src="https://github.com/user-attachments/assets/50baf24d-aee5-44d9-9fd4-9699c3899548" />

Onboard in Gentoo (running under Xwayland) now works as before, with Onboard-Settings->Keyboard->Advanced->Input Event Source as XInput.  'Key-stroke generator' works on any setting for me, except for AT-SPI.  This includes Auto, Xtest and Uinput.

**Comment by dr-ni** [comment]
please check onboard now without xwayland with the latest merge #49

**Comment by rab0171610** [comment]
First, as I said, using under Xwayland is actually ideal for now IMO.  I think Wayland addresses security issues that Redhat and corporate Linux is more concerned about than the average desktop user.  We have had the same X11 level security on single user desktop and laptops for over 30 years now and most of us have not experienced any security breaches because of it. 

That being said, I did try the update in Gentoo under Wayland.  I experienced some glitches including visually with the keyboard interface (a noticeable outline/window border around the edge of the keyboard, red feedback keys when typing even when CapsLock not engaged . . ..  It would not dock into the KDE System Tray and there was no tray icon.  Essentially, it was not usable.  It may be a problem on my end.  If and when I get more time to troubleshoot it, I will try again.  For the now, I am going to continue using an older release under Xwayland for stability.  
On subsequent attempts to test the new version under Wayland, I will engage by providing feedback and opening issues if needed.
For now, I think the information I provided above is important to maintain a working keyboard.  Many people may not have the option to build onboard from source and may be stuck with older versions of onboard that are found in their distro's repository and may not see the most recent version for some months to some years.  So it is important that all options that keep older releases of onboard working in Wayland under these various conditions are documented.  

As of today, I am not able to provide any meaningful feedback but I will do so in the future.

**Comment by AlexP11223** [comment]
> For the now, I am going to continue using an older release under Xwayland for stability.

You should be able to do the same in the new release too, if you add `GDK_BACKEND=x11` it will not use the new Wayland approach. 

> It would not dock into the KDE System Tray and there was no tray icon.

Yes, docking currently does not work. It is strange, however, that the tray icon disappeared (I reproduced it too).

> First, using 'GDK_BACKEND=x11 onboard' is an acceptable workaround for me. For my personal use cases on my personal desktops, I do not feel that additional security measures in Wayland are addressing problems that I have. X11 level security is acceptable to me.

The main problem was that it did not work reliably, depending on the compositor etc.

**Comment by rab0171610** [comment]
> 
> You should be able to do the same in the new release too, if you add `GDK_BACKEND=x11` it will not use the new Wayland approach.
> 

That may be, but there I have no pressing need to migrate to the newer version.    
 
> Yes, docking currently does not work. It is strange, however, that the tray icon disappeared (I reproduced it too).

That is a regression, a major bug and a showstopper.  This is all the more reason to stick to a working prior release.  I only use Onboard by clicking on the tray icon as needed.  I do not want it auto appearing just because an type-able field is in focus.  The systray icon is a feature that I heavily utilize and require.   
 
> The main problem was that it did not work reliably, depending on the compositor etc.
It works reliably in KDE Plasma as of now.  As I don't plan to use another compositor besides Kwin in KDE so that does not affect me.
It also works reliably in Gnome.  It also works in X11 (XFCE, Mate, Cinnamon) That means it can work in a majority of the likely use cases. So it may not work in every compositor. If working in every compositor versus working in the majority of the real world use cases is a goal of  the current development of Onboard then that is lofty at best, especially when the current release is missing important features.  Maybe narrow the current focus to just getting it to work in Wayland without losing major features.      

I appreciate the effort in getting Onboard to work reliably in Wayland.  However, it is clearly still a work in progress.  I will continue to use the prior version in Gentoo and the one available in the Debian repository respectively for the interim, both are working as expected.  I will be happy to test the newer releases accordingly as work progresses but it is not an emergency.   When I have time, to support this effort , I will work on building, installing, testing the most recent versions.  Until then, the prior version is working as expected under XWayland in two different distros, at least for me.  In one scenario, Debian I am using the uinput method.  In Gentoo, using the newest versions of KDE requires a window rule to workaround a change in Kwin, the compositor in order to prevent Onboard from getting focus when trying to type instead of the application. In both, Xwayland is still required.  Other users may also find this more reliable than migrating to a newer version just to avoid using Xwayland.  I guess it depends on your priorities.

**Comment by AlexP11223** [comment]
> That is a regression, a major bug and a showstopper.

I mean, it seems to happen only when using the Wayland mode. If you added `GDK_BACKEND=x11`, then nothing should change. 

Personally, I have never used the icons, I simply added it into the taskbar.

**Comment by rab0171610** [comment]
> > That is a regression, a major bug and a showstopper.
> 
> I mean, it seems to happen only when using the Wayland mode. If you added `GDK_BACKEND=x11`, then nothing should change.
> 
> Personally, I have never used the icons, I simply added it into the taskbar.

As I said, I will test that at some point and report back.  I start Onboard hidden so that it is not in the way after logging in so It silently docks into the systray.  I use the physical keyboard primarily but when I need to use Onboard instead, I summon it by clicking on the System Tray icon.   Clicking on the same icon docks it away again.  When they systray icon functionality is not working, such as in Wayland in the newest release, it instead appears as an indicator in the Icons-only Task Manager, which I assume is what you are calling the taskbar.  I already have enough applications pinned there, browsers, file managers, System Settings, Email, etc.   I do not wish to add any more.   I could pin it there but that is not actually the same functionality as Onboard already running and discretely sitting in the System Tray ready to use as needed.  It is a much smaller icon also takes up much less screen real estate that way.

**Comment by AlexP11223** [comment]
Actually, I can't reproduce the tray icon issue anymore, so it's either random, or it got fixed by the latest KDE update I installed yesterday after that, or I somehow got confused because normally I do not use this feature.

Right now it seems to work fine for me, the only issue is that after restoring the Onboard window goes to the default position (center), but it is a general Wayland issue for all apps.

**Comment by rab0171610** [comment]
Yeah, NO.  Whose bright idea was it to change the color scheme of pressed keys in the Nightshade theme to a from a darker blue/gray to a BRIGHT RED, a color previously reserved for letting you know the caps lock key was engaged in onboard?  Terrible.  If you are going to mess with long established themes at least code a new configuration option to toggle this color.  There do not appear to be any bugs or issues opened requesting this change. This appears to be an accidental bug or someone's arbitrary design decision.  I guess I will fork this project and revert the commit that enabled this. This is not an improvement.

**Comment by dr-ni** [comment]
With the introduction of key hover effects this was globally set for all schemes

**Comment by AlexP11223** [comment]
> There do not appear to be any bugs or issues opened requesting this change. 

Then maybe you should create it. :)

But yeah, I am also not a fan of sudden changes in the themes. This one, though, seems quite minor and did not bother me/did not notice it. I am using a bit simplified Nightshade (via the customization window): [cat ~/.local/share/onboard/themes/Nightshade.theme](https://gist.github.com/AlexP11223/f211b9e53b95888e2b1007d50d3f549b)

**Comment by rab0171610** [comment]
> With the introduction of key hover effects this was globally set for all schemes

Yeah well I have forked it and will just undo it.  That is visually just awful and a heavy handed design decision..

Apparently this effort to enable Onboard to work in Wayland has become a playground for someone to tinker with the code.  Stark arbitrary visual changes to themes that people have used for years is not acceptable.  This creates a break in the UI from one version to the next.

**Comment by rab0171610** [comment]
> With the introduction of key hover effects this was globally set for all schemes

Please explain the necessity for "key hover effects"  and what they are intended to accomplish exactly.  I do not see any mention of this in the README or any documented plans or published discussions about why this is being implemented. The last we heard is that the project was looking for help getting onboard to work with Wayland and asking for ideas.  Now this.  The only references to Hover effects would be if I read through the commit changes or your comment here.

**Comment by AlexP11223** [comment]
> Please explain the necessity for "key hover effects" and what they are intended to accomplish exactly.

It was my request :) https://github.com/onboard-osk/onboard/issues/12
And yes, it is not related to Wayland, so it should be discussed somewhere else.

**Comment by rab0171610** [comment]
Well your requested changes break every theme and force a non-theme key press color in user's respective themes.   I will politely remind you that this is an accessibility utility program used by many people with disabilities.  Bright red is very stark when it does not follow the traditional system theme, especially for people with vision deficiencies.  Red is traditionally used a a warning color indicating breakage or an error.  This is even more significant when you consider that the bright or dark red was traditionally used as a warning color in Onboard to indicate to the user that the Caps Lock Key was engaged.  Enjoy your hover effects.

**Comment by dr-ni** [comment]
The reason for color change was simply the problem that hover effect and click effect were rather the same - and confusing

Hover effects are a significant accessibility enhancemet - I love it
Feel free to make your own theme

**Comment by AlexP11223** [comment]
> Well your requested changes break every theme and force a non-theme key press color in user's respective themes.

Well, I think I was not requesting any specific colors. It could be that red is indeed not the best choice, iirc the keyboards on Windows use some kind of white, grey colors for pressed/hover. And ideally it should be possible to customize the colors for these effects, afaik currently it is not possible, at least not via the UI or user files.

But I think you should create a separate issue for that if you have some specific improvement suggestions, this is not related to Wayland.

**Comment by gnu-like** [comment]
Really Thanks,finally I could run onboard on debian kde plasma wayland,
1- onboard icon top panel missing
2- suggestions not working

**Comment by AlexP11223** [comment]
What do you mean by "top panel"? The window decorations? 🤔  Seem to work fine for me.

<img width="747" height="324" alt="Image" src="https://github.com/user-attachments/assets/4b5f0794-278c-4576-af15-7da53be53aa6" />

Suggestions like suggested words? I just tried to enable that and it seems to work for me, tried it in Firefox and Kate. But if using multiple languages, then there is the same issue as with snippets that was mentioned in [README.WAYLAND.md](https://github.com/onboard-osk/onboard/blob/main/README.WAYLAND.md).

**Comment by gnu-like** [comment]
top of the desktop beside clock system tray , we click onbaord icon minimize and show

**Comment by dr-ni** [comment]
since the main adaption has now almost been done now I will close this
